crates/core/src/store.rs

```diff
@@ -336,29 +288,42 @@ impl StoreBuilder {
         host_path: impl AsRef<Path>,
         guest_path: PathBuf,
     ) -> Result<()> {
-        let dir =
-            || cap_std::fs::Dir::open_ambient_dir(host_path.as_ref(), cap_std::ambient_authority());
+        self.preopened_dir_impl(host_path, guest_path, true)
+    }
+
+    fn preopened_dir_impl(
+        &mut self,
+        host_path: impl AsRef<Path>,
+        guest_path: PathBuf,
+        writable: bool,
+    ) -> Result<()> {
+        let cap_std_dir =
+            cap_std::fs::Dir::open_ambient_dir(host_path.as_ref(), cap_std::ambient_authority())?;
         let path = guest_path
             .to_str()
             .ok_or_else(|| anyhow!("non-utf8 path: {}", guest_path.display()))?;
 
         self.try_with_wasi(|wasi| {
             match wasi {
-                Wasi::Preview1(ctx) => ctx.push_preopened_dir(
-                    Box::new(wasmtime_wasi_preview1::dir::Dir::from_cap_std(dir()?)),
-                    path,
-                )?,
-                Wasi::Preview2(ctx) => ctx.push_preopened_dir(
-                    Box::new(wasmtime_wasi_preview2::dir::Dir::from_cap_std(dir()?)),
-                    path,
-                )?,
+                Wasi::Preview1(ctx) => {
+                    let mut dir =
+                        Box::new(wasmtime_wasi_preview1::dir::Dir::from_cap_std(cap_std_dir)) as _;
+                    if !writable {
+                        dir = Box::new(preview1::ReadOnlyDir(dir));
+                    }
+                    ctx.push_preopened_dir(dir, path)?;
+                }
+                Wasi::Preview2(ctx) => {
+                    let mut dir =
+                        Box::new(wasmtime_wasi_preview2::dir::Dir::from_cap_std(cap_std_dir)) as _;
+                    if !writable {
+                        dir = Box::new(wasi_preview2::dir::ReadOnlyDir(dir));
+                    }
+                    ctx.push_preopened_dir(dir, path)?;
```

Comment:
nit: can this be moved out of the match statement to make it clear that that's what's happening regardless of the which wasi preview.

Reply:
Ah sorry! I read them as both trait objects of the same type. Whoops!